CUDADataFormats/TrackingRecHit/src/TrackingRecHit2DHeterogeneous.cc

```diff
@@ -4,38 +4,83 @@
+  auto ret = cms::cuda::make_host_unique<float[]>(static_cast<int>(this->n32) * this->nHits(), stream);
+  cms::cuda::copyAsync(ret, this->m_store32, static_cast<int>(this->n32) * this->nHits(), stream);
   return ret;
 }
 
-template <>
-cms::cuda::host::unique_ptr<uint16_t[]> TrackingRecHit2DGPU::store16ToHostAsync(cudaStream_t stream) const {
-  auto ret = cms::cuda::make_host_unique<uint16_t[]>(static_cast<int>(n16) * nHits(), stream);
-  cms::cuda::copyAsync(ret, m_store16, static_cast<int>(n16) * nHits(), stream);
+template <typename TrackerTraits>
+cms::cuda::host::unique_ptr<uint16_t[]> TrackingRecHit2DGPUT<TrackerTraits>::store16ToHostAsync(
+    cudaStream_t stream) const {
+  auto ret = cms::cuda::make_host_unique<uint16_t[]>(static_cast<int>(this->n16) * this->nHits(), stream);
+  cms::cuda::copyAsync(ret, this->m_store16, static_cast<int>(this->n16) * this->nHits(), stream);
   return ret;
 }
 
-template <>
-cms::cuda::host::unique_ptr<uint32_t[]> TrackingRecHit2DGPU::hitsModuleStartToHostAsync(cudaStream_t stream) const {
-  auto ret = cms::cuda::make_host_unique<uint32_t[]>(nMaxModules() + 1, stream);
-  cudaCheck(
-      cudaMemcpyAsync(ret.get(), m_hitsModuleStart, sizeof(uint32_t) * (nMaxModules() + 1), cudaMemcpyDefault, stream));
+template <typename TrackerTraits>
+cms::cuda::host::unique_ptr<uint32_t[]> TrackingRecHit2DGPUT<TrackerTraits>::hitsModuleStartToHostAsync(
+    cudaStream_t stream) const {
+  auto ret = cms::cuda::make_host_unique<uint32_t[]>(TrackerTraits::numberOfModules + 1, stream);
+  cudaCheck(cudaMemcpyAsync(ret.get(),
+                            this->m_hitsModuleStart,
+                            sizeof(uint32_t) * (TrackerTraits::numberOfModules + 1),
+                            cudaMemcpyDefault,
+                            stream));
   return ret;
 }
 
-// the only specialization needed
-template <>
-void TrackingRecHit2DHost::copyFromGPU(TrackingRecHit2DGPU const* input, cudaStream_t stream) {
+//NB these specialization below from the base class seems to be needed since they are used by the constructor (that is in the base class)
+// and if not explicilty defined the linker doesn't see them. FIXME: is there a cleaner way to do this that would still allow "partial specialization"?
+// With alpaka or the unique memory pool I expect this issues to be gone
+
+template <typename Traits, typename TrackerTraits>
+cms::cuda::host::unique_ptr<float[]> TrackingRecHit2DHeterogeneousT<Traits, TrackerTraits>::localCoordToHostAsync(
+    cudaStream_t stream) const {
+  if constexpr (std::is_same_v<Traits, cms::cudacompat::GPUTraits>) {
+    auto ret = cms::cuda::make_host_unique<float[]>(5 * this->nHits(), stream);
+    cms::cuda::copyAsync(ret, this->m_store32, 5 * this->nHits(), stream);
+    return ret;
+  } else {
+    static_assert(true, "Intended to be used only with GPU traits.\n");
+    return nullptr;
+  }
+}
```

Comment:
This may no longer be needed after changing the way inheritance is used.
Could you check ?

Reply:
Sure, let me check it

Reply:
Given the simplicity of these functions, it should be easy enough to just inline them in the construcotr itself and get rid of all the extra complexity and code duplication:
```c++
template <typename Traits, typename TrackerTraits>
TrackingRecHit2DHeterogeneousT<Traits, TrackerTraits>::TrackingRecHit2DHeterogeneousT(
    uint32_t nHits,
    int32_t offsetBPIX2,
    pixelCPEforGPU::ParamsOnGPUT<TrackerTraits> const* cpeParams,
    uint32_t const* hitsModuleStart,
    cudaStream_t stream,
    TrackingRecHit2DHeterogeneousT<cms::cudacompat::GPUTraits, TrackerTraits> const* input)
    : m_nHits(nHits), m_offsetBPIX2(offsetBPIX2), m_hitsModuleStart(hitsModuleStart) {
  using TrackingRecHit2DSOAView = TrackingRecHit2DSOAViewT<TrackerTraits>;

  auto view = Traits::template make_host_unique<TrackingRecHit2DSOAView>(stream);

  view->m_nHits = nHits;
  m_view = Traits::template make_unique<TrackingRecHit2DSOAView>(stream);  // leave it on host and pass it by value?
  m_AverageGeometryStore = Traits::template make_unique<typename TrackingRecHit2DSOAView::AverageGeometry>(stream);
  view->m_averageGeometry = m_AverageGeometryStore.get();
  view->m_cpeParams = cpeParams;
  view->m_hitsModuleStart = hitsModuleStart;

  // if empty do not bother
  if (0 == nHits) {
    if constexpr (std::is_same_v<Traits, cms::cudacompat::GPUTraits>) {
      cms::cuda::copyAsync(m_view, view, stream);
    } else {
      m_view.reset(view.release());  // NOLINT: std::move() breaks CUDA version
    }
    return;
  }

  // the single arrays are not 128 bit alligned...
  // the hits are actually accessed in order only in building
  // if ordering is relevant they may have to be stored phi-ordered by layer or so
  // this will break 1to1 correspondence with cluster and module locality
  // so unless proven VERY inefficient we keep it ordered as generated

  // host copy is "reduced"  (to be reviewed at some point)
  if constexpr (std::is_same_v<Traits, cms::cudacompat::HostTraits>) {
    assert(input != nullptr);

    this->m_store32 = cms::cuda::make_host_unique<float[]>(5 * input->nHits(), stream);
    cms::cuda::copyAsync(this->m_store32, input->m_store32, 5 * input->nHits(), stream);
  } else {
    assert(input == nullptr);

    m_store16 = Traits::template make_unique<uint16_t[]>(nHits * n16, stream);
    m_store32 = Traits::template make_unique<float[]>(nHits * n32 + TrackerTraits::numberOfLayers + 1, stream);
    m_PhiBinnerStore = Traits::template make_unique<typename TrackingRecHit2DSOAView::PhiBinner>(stream);
  }

  static_assert(sizeof(typename TrackingRecHit2DSOAView::hindex_type) == sizeof(float));
  static_assert(sizeof(typename TrackingRecHit2DSOAView::hindex_type) ==
                sizeof(typename TrackingRecHit2DSOAView::PhiBinner::index_type));

  auto get32 = [&](Storage32 i) { return m_store32.get() + static_cast<int>(i) * nHits; };

  // copy all the pointers
  m_phiBinner = view->m_phiBinner = m_PhiBinnerStore.get();
  m_phiBinnerStorage = view->m_phiBinnerStorage =

      reinterpret_cast<typename TrackingRecHit2DSOAView::PhiBinner::index_type*>(get32(Storage32::kPhiStorage));

  view->m_xl = get32(Storage32::kXLocal);
  view->m_yl = get32(Storage32::kYLocal);
  view->m_xerr = get32(Storage32::kXerror);
  view->m_yerr = get32(Storage32::kYerror);
  view->m_chargeAndStatus = reinterpret_cast<uint32_t*>(get32(Storage32::kCharge));

  if constexpr (not std::is_same_v<Traits, cms::cudacompat::HostTraits>) {
    view->m_xg = get32(Storage32::kXGlobal);
    view->m_yg = get32(Storage32::kYGlobal);
    view->m_zg = get32(Storage32::kZGlobal);
    view->m_rg = get32(Storage32::kRGlobal);

    auto get16 = [&](Storage16 i) { return m_store16.get() + static_cast<int>(i) * nHits; };
    m_iphi = view->m_iphi = reinterpret_cast<int16_t*>(get16(Storage16::kPhi));

    view->m_xsize = reinterpret_cast<int16_t*>(get16(Storage16::kXSize));
    view->m_ysize = reinterpret_cast<int16_t*>(get16(Storage16::kYSize));
    view->m_detInd = get16(Storage16::kDetId);

    m_phiBinner = view->m_phiBinner = m_PhiBinnerStore.get();
    m_hitsLayerStart = view->m_hitsLayerStart = reinterpret_cast<uint32_t*>(get32(Storage32::kLayers));
  }

  // transfer view
  if constexpr (std::is_same_v<Traits, cms::cudacompat::GPUTraits>) {
    cms::cuda::copyAsync(m_view, view, stream);
  } else {
    m_view.reset(view.release());  // NOLINT: std::move() breaks CUDA version
  }
}
```
though it needs to declare the different specialisations as friends:
```c++
template <typename Traits, typename TrackerTraits>
class TrackingRecHit2DHeterogeneousT {
  ...
  template <typename T, typename U>
  friend class TrackingRecHit2DHeterogeneousT;
  ...
```

Reply:
Mmm, I would say that cpu and gpu share most of the code (everything excepted view transferring) while host is basically another constructor ( see the `assert(input == nullptr);` for non host) so something in between could work. Something like having the common constructor for  cpu/gpu with the inlined functions above and one fully specialized for host (that is nearer to a "copy" constructor). I'll try to figure something out!

Reply:
I've resorted the constructors separating the one for the Host and having it only in the specialization and having the ones for cpu and gpu deriving from base. Let me know what you think about it. I still need to fix the `unique_ptr` below.